### PR TITLE
Add junit saltcheck

### LINKED ISSUE #63463: [FEATURE REQUEST] JUnit output for saltcheck
**Describe the solution you'd like**
I'd really love it if saltcheck could output in JUnit so my CI could interpret the results.

**Describe alternatives you've considered**
Parse the results outside of salt.

**Please Note**
If this feature request would be considered a substantial change or addition, this should go through a SEP process here https://github.com/saltstack/salt-enhancement-proposals, instead of a feature request.

### PR DESCRIPTION
**Pull request by M3GH4NN**
### What does this PR do?
adds JUnit output to saltcheck

### What issues does this PR fix or reference?
Fixes: #63463 

### Previous Behavior
Could not output JUnit

### New Behavior
Now you can output JUnit with saltcheck

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [X] Docs
- [X] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [X] Tests written/updated

### Commits signed with GPG?
No

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: requirements/static/ci/common.in [hunk 1]
```diff
@@ -18,6 +18,7 @@ etcd3-py==0.1.6 ; python_version >= '3.6'
 gitpython>=3.1.30; python_version >= '3.7'
 jmespath
 jsonschema
+junit_xml>=1.9
```
Comment (twangboy) [inline requirements/static/ci/common.in:21]: Shouldn't this be  `junit-xml`?

**Review by twangboy — CHANGES_REQUESTED** [review]

File: salt/modules/saltcheck.py [hunk 2]
```diff
@@ -317,6 +317,14 @@
 from salt.utils.json import dumps, loads
 from salt.utils.odict import OrderedDict
 
+try:
+    from junit_xml import TestCase, TestSuite
+
+    HAS_JUNIT = True
+except (ImportError, ModuleNotFoundError):
```
Comment (s0undt3ch) [inline salt/modules/saltcheck.py:324]: Why would you get `ModuleNotFound` and not `ImportError`?

File: requirements/static/ci/common.in (hunk 1 above)
Comment (s0undt3ch) [reply requirements/static/ci/common.in:21]: Yes, please change to what @twangboy said. Even if its named with an underscore on PyPi.

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

File: salt/modules/saltcheck.py (hunk 2 above)
Comment (M3GH4NN) [reply salt/modules/saltcheck.py]: Fixed this

**Review by s0undt3ch — DISMISSED** [review]

**Comment by s0undt3ch** [comment]
Test failure looks related.

File: tests/pytests/unit/modules/test_saltcheck.py [hunk 3]
```diff
@@ -3,6 +3,11 @@
 import salt.modules.saltcheck as saltcheck
 from tests.support.mock import MagicMock
 
+try:
+    from xmldiff import main
+except ImportError:
+    raise ImportError
+
```
Comment (s0undt3ch) [inline tests/pytests/unit/modules/test_saltcheck.py:10]:
```suggestion
xmldiff = pytest.importorskip("xmldiff")
```

File: tests/pytests/unit/modules/test_saltcheck.py [hunk 4]
```diff
@@ -13,3 +18,29 @@ def configure_loader_modules():
...
+        """<?xml version="1.0" ?>\n<testsuites disabled="0" errors="0" failures="1" tests="3" time="3.041">\n"""
+        + """\t<testsuite disabled="0" errors="0" failures="1" name="test_results" skipped="1" tests="3" time="3.041">\n"""
+        + """\t\t<testcase name="echo_test_hello" time="2.590700"/>\n"""
+        + """\t\t<testcase name="echo_test_hello2" time="0.450300">\n"""
+        + """\t\t\t<failure type="failure" message="Fail: fail expected is not equal to hello"/>\n\t\t</testcase>\n"""
+        + """\t\t<testcase name="echo_test_hello3">\n\t\t\t<skipped type="skipped" message="Skip"/>\n\t\t</testcase>\n"""
+        + """\t</testsuite>\n</testsuites>\n"""
+    )
+    ret = saltcheck._generate_junit_out_list(results)
+    diff = main.diff_texts(ret, expected)
```
Comment (s0undt3ch) [inline tests/pytests/unit/modules/test_saltcheck.py:45]:
```suggestion
    diff = xmldiff.main.diff_texts(ret, expected)
```

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

**Review by s0undt3ch — DISMISSED** [review]

**Comment by whytewolf** [comment]
looks like precommit failed. for black and pylint.

File: salt/modules/saltcheck.py [hunk 5]
```diff
@@ -517,7 +527,11 @@ def run_state_tests(state, saltenv=None, check_all=False, only_fails=False):
         # If passed a duplicate state, don't overwrite with empty res
         if not results.get(state_name):
             results[state_name] = results_dict
-    return _generate_out_list(results, only_fails=only_fails)
+
+        if junit and HAS_JUNIT:
+            return _generate_junit_out_list(results)
+        else:
+            return _generate_out_list(results, only_fails=False)
```
Comment (nicholasmhughes) [inline salt/modules/saltcheck.py:534]: **issue (blocking):** I think this was set to `False` during testing and made its may into the PR...

```suggestion
            return _generate_out_list(results, only_fails=only_fails)
```

**Review by nicholasmhughes — CHANGES_REQUESTED** [review]

**Comment by nicholasmhughes** [comment]
re-run pr-arch-lts-x86_64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-centos-7-x86_64-py3-pytest

**Review by nicholasmhughes — DISMISSED** [review]

**Comment by nicholasmhughes** [comment]
re-run pr-photon-3-x86_64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-centos-7-x86_64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-photon-3-x86_64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-arch-lts-x86_64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-photon-3-x86_64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-photon-3-x86_64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-photon-3-x86_64-py3-pytest

**Review by whytewolf — DISMISSED** [review]

**Review by twangboy — DISMISSED** [review]

File: salt/modules/saltcheck.py [hunk 6]
```diff
@@ -442,6 +452,7 @@ def run_state_tests(state, saltenv=None, check_all=False, only_fails=False):
     :param str saltenv: optional saltenv. Defaults to base
     :param bool check_all: boolean to run all tests in state/saltcheck-tests directory
     :param bool only_fails: boolean to only print failure results
+    :param bool junit: boolean to print results in junit format
```
Comment (Ch3LL) [inline salt/modules/saltcheck.py:455]:
```suggestion
    :param bool junit: boolean to print results in junit format
        .. versionadded:: 3007.0
```

File: salt/modules/saltcheck.py [hunk 7]
```diff
@@ -534,12 +548,13 @@ def parallel_scheck(data):
...
 
 
-def run_highstate_tests(saltenv=None, only_fails=False):
+def run_highstate_tests(saltenv=None, only_fails=False, junit=False):
     """
     Execute all tests for states assigned to the minion through highstate and return results
 
     :param str saltenv: optional saltenv. Defaults to base
     :param bool only_fails: boolean to only print failure results
+    :param bool junit: boolean to print results in junit format
```
Comment (Ch3LL) [inline salt/modules/saltcheck.py:557]:
```suggestion
    :param bool junit: boolean to print results in junit format
        .. versionadded:: 3007.0
```

**Review by Ch3LL — CHANGES_REQUESTED** [review]

File: salt/modules/saltcheck.py [hunk 8]
```diff
@@ -618,6 +635,35 @@ def _generate_out_list(results, only_fails=False):
...
+                status = val["status"]
+                test_cases.append(TestCase(name, "", round(time, 4)))
+                if status.startswith("Fail"):
+                    failed = 1
+                    test_cases[len(test_cases) - 1].add_failure_info(status)
+                if status.startswith("Skip"):
+                    test_cases[len(test_cases) - 1].add_skipped_info(status)
+                total_time = total_time + float(val["duration"])
+    test_suite = TestSuite("test_results", test_cases)
+    # Set exist code to 1 if failed tests
```
Comment (nicholasmhughes) [inline salt/modules/saltcheck.py:660]: **nitpick (non-blocking):** `exist` > `exit`

```suggestion
    # Set exit code to 1 if failed tests
```

**Review by Ch3LL — DISMISSED** [review]

**Comment by Ch3LL** [comment]
Looks like pre-commit is failing

**Comment by Ch3LL** [comment]
ping @s0undt3ch can you re-review?

**Comment by nicholasmhughes** [comment]
@s0undt3ch can haz review plz?  :pleading_face:

**Review by s0undt3ch — DISMISSED** [review]

**Comment by Ch3LL** [comment]
Apologies, but looks like a merge conflict that needs to be resolved here. Let me know if I can help here

**Comment by Ch3LL** [comment]
I went ahead and rebased and fixed some new issues with pre-commit. @nicholasmhughes you mind reviewing to ensure I didn't mess anything up?

**Review by nicholasmhughes — APPROVED** [review]

**Comment by nicholasmhughes** [comment]
@Ch3LL , looks good!

**Review by Ch3LL — APPROVED** [review]

**Pull request merged**